[TwigBridge] Added template "name" argument to debug:twig command to find their paths

yceruto opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #27911
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/10232

Find the template file (to load by Twig) from a given template name (useful to know which file will be loaded exactly and which ones don't):
![debug-twig-loader-overridden](https://user-images.githubusercontent.com/2028198/42849959-81a8c49a-89f3-11e8-8d93-21581fe606a9.png)
This will also show the overridden files if they exist and the paths corresponding to their namespace.

In addition, the command suggests alternatives if you made a typo (this way you can check your template name quickly):

| namespace typo | template name typo |
| --- | --- | 
| ![debug-twig-loader-ns-typo-alt](https://user-images.githubusercontent.com/2028198/42850624-81803e3c-89f6-11e8-8a92-11f09c99d13c.png) | ![debug-twig-loader-typo-alt](https://user-images.githubusercontent.com/2028198/42850644-99571238-89f6-11e8-9cf7-ed9b880f3d81.png) |

<details>
<summary><strong>Other outputs</strong></summary>

Discovering more alternatives:
![debug-twig-loader-not-found-many-alt](https://user-images.githubusercontent.com/2028198/42850815-82a30eb0-89f7-11e8-8d23-530f8ff325bc.png)

Unknown template name:
![debug-twig-loader-not-found](https://user-images.githubusercontent.com/2028198/42850882-d647aad0-89f7-11e8-9735-94149895437f.png)
</details>


## Update
The feature was introduced into `debug:twig` command and the `filter` argument was converted to `--filter` option. The `name` argument is now the first one of the command.

yceruto committed:
[TwigBridge] Added template \"name\" argument to debug:twig command to find their paths

nicolas-grekas commented:
Should it be a dedicated command or an option on the `debug:twig` one?
I'm not sure growing the list of commands with this one is justified (it's a "minor" command, isn't it?)

yceruto commented:
IMHO the `name` argument doesn't seem like an option, but an argument that represents a pure action (find the path of the template name). 

At first I thought about adding it to `debug:twig` but this already has a main argument (`filter`) for the general info of Twig (functions, filters, tests, globals), I thought that adding this new feature would be a mess or it would require a complete refactor of the command (including passing the argument `filter` to option ` --filter` and making `name` as the main argument of the command).

I'd also like to have everything in one command, but its initial design should change then. I imagine something like this:
```bash
$ b/c debug:twig # show general info

# proposal 1 (without refactor)
$ b/c debug:twig escape # filter the general info
$ b/c debug:twig --find=@Twig/Exception/error.html.twig # find the template path

# proposal 2 (with refactor) much better to me.
$ b/c debug:twig @Twig/Exception/error.html.twig # find the template path
$ b/c debug:twig --filter=escape # filter the general info
```
Do you prefer this?

yceruto commented:
@nicolas-grekas I added a second commit with your suggestion and `# proposal 2`. I also added a missing (minor) feature.

 MINOR FEATURE:
 * allowed `--filter` with `--format=json`

yceruto commented:
Fixing AppVeyor failure...

Status: Needs Work

yceruto commented:
(Travis failures unrelated)

Status: Needs Review

fabpot commented on src/Symfony/Bridge/Twig/Command/DebugCommand.php:
```diff
@@ -12,11 +12,13 @@
 use Symfony\Component\Console\Input\InputOption;
 use Symfony\Component\Console\Output\OutputInterface;
 use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Finder\Finder;
```
Finder is currently not a dep of Twig bridge.

yceruto replied:
All dependencies, including `symfony/finder` are in `require-dev` section:
https://github.com/symfony/symfony/blob/b67c26c2efe4673553d6379367ebecc2eac7dca5/src/Symfony/Bridge/Twig/composer.json#L23-L30

Note that the `Finder` component was already used in other places like `TwigExtractor` and `LintCommand`, do we need to do something else?

fabpot commented on src/Symfony/Bridge/Twig/Command/DebugCommand.php:
```diff
@@ -50,19 +52,24 @@ protected function configure()
             ->setDefinition(array(
-                new InputArgument('filter', InputArgument::OPTIONAL, 'Show details for all entries matching this filter'),
+                new InputArgument('name', InputArgument::OPTIONAL, 'The template name'),
+                new InputOption('filter', null, InputOption::VALUE_REQUIRED, 'Show details for all entries matching this filter'),
```
This needs to be documented in the CHANGELOG

yceruto replied:
Updated CHANGELOG.

fabpot commented on src/Symfony/Bridge/Twig/Command/DebugCommand.php:
```diff
@@ -50,19 +52,24 @@ protected function configure()
+
+The command lists all paths that match the given template name.
+
+  <info>php %command.full_name% --filter=date</info>
```
This changes the command completely. Not sure if we want to keep BC on commands (I think we do not), but that means we need a PR on docs as well.

yceruto replied:
I'll do it as soon as possible.

fabpot commented:
Thank you @yceruto.
